shadowsocks/tcp.go

```diff
@@ -55,18 +55,27 @@ func remoteIP(conn net.Conn) net.IP {
 	if tcpaddr, ok := addr.(*net.TCPAddr); ok {
 		return tcpaddr.IP
 	}
-	ipstr, _, err := net.SplitHostPort(conn.RemoteAddr().String())
+	ipstr, _, err := net.SplitHostPort(addr.String())
 	if err == nil {
 		return net.ParseIP(ipstr)
 	}
 	return nil
 }
 
+// Wrapper for logger.Debugf with a fixed number of arguments.
+// This is an optimization to reduce unnecessary allocations due to an interaction
+// between Go's inlining/escape analysis and varargs functions like logger.Debugf.
+func debugLogID(template, id string, v2 interface{}) {
```

Comment:
Let's make this TCP-specific and create another one for UDP. I'd like for the TCP and UDP codes to not depend on each other.

Reply:
Done!